### DISCUSSION TITLE
Selecting a residue

### DISCUSSION THREAD

**Discussion by Manubi** [molstar/molstar#998] [Q&A] [answered]
Hi,

this is how my current component looks like. The idea is to click a row in the table. I parse the residue number i pass it to the MolstarWrapper which the code is below and then to highlight the selected residue from the table. Or is there another way to select the correct residue from the table in the Molstar protein?

Thanks for your help!

![CleanShot 2023-12-05 at 16 10 14](https://github.com/molstar/molstar/assets/20935138/423d0b1d-c753-4107-8047-59967a29f14d)

```ts
import { createPluginUI } from "molstar/lib/mol-plugin-ui";
import { PluginUIContext } from "molstar/lib/mol-plugin-ui/context";
import { createRef, useEffect } from "react";

import { Structure } from "molstar/lib/mol-model/structure";
import "molstar/lib/mol-plugin-ui/skin/light.scss";

declare global {
  interface Window {
    molstar?: PluginUIContext;
  }
}

export function MolStarWrapper({
  url,
  selectedResidue,
}: {
  url: string;
  selectedResidue: number;
}) {
  const parent = createRef<HTMLDivElement>();

  useEffect(() => {
    async function init() {
      window.molstar = await createPluginUI(parent.current as HTMLDivElement);

      const data = await window.molstar.builders.data.download(
        {
          url,
        },
        { state: { isGhost: true } }
      );
      const trajectory =
        await window.molstar.builders.structure.parseTrajectory(data, "pdb");
      await window.molstar.builders.structure.hierarchy.applyPreset(
        trajectory,
        "default"
      );
    }
    init();
    return () => {
      window.molstar?.dispose();
      window.molstar = undefined;
    };
  }, []);

  useEffect(() => {
// selectedResdue == some number eg. 31, 204... 
// here i need to select the residue and highlight it in the window.molstar
  }, [selectedResidue]);

  return (
    <div className="relative">
      <div ref={parent} style={{ width: 640, height: 640 }} />
    </div>
  );
}
```

**Comment by dsehnal** [comment]
The basic wrapper example shows how to highlight things based on events:
https://github.com/molstar/molstar/blob/master/src/examples/basic-wrapper/index.ts#L141
https://github.com/molstar/molstar/blob/master/src/examples/basic-wrapper/index.html#L105

**Reply by Manubi** [reply]
is there also the possibility to paint the highlighted thing in a different color? because not the problem is as soon as i move from the table onto molstar the selected thing disappears.

**Reply by dsehnal** [reply]
You can use `select` instead of `highlight` or the "overpaint" mechanism.

**Reply by Manubi** [reply]
This is what i have currenlty which works fine for the highlighting. Can't really wrap my head around how the overpaint or select could work.
Do you have another hint? thanks!

```ts
async function selectResidue(selectedResidue: number) {
    const data =
      window.molstar?.managers.structure.hierarchy.current.structures[0]?.cell
        .obj?.data;
    if (!data) return;

    const seq_id = selectedResidue;
    const sel = Script.getStructureSelection(
      (Q) =>
        Q.struct.generator.atomGroups({
          "residue-test": Q.core.rel.eq([
            Q.struct.atomProperty.macromolecular.label_seq_id(),
            seq_id,
          ]),
          "group-by": Q.struct.atomProperty.macromolecular.residueKey(),
        }),
      data
    );
    const loci = StructureSelection.toLociWithSourceUnits(sel);
    window.molstar?.managers.interactivity.lociHighlights.highlightOnly({
      loci,
    });
  }
```

**Reply by dsehnal** [reply, marked as answer]
- For select, `interactivity.lociHighlights` => `interactivity.lociSelects`
  - You can tweak the select color using `plugin.canvas3d?.setProps`, although this will always be global
- For overpaint, you can use [this function](https://github.com/molstar/molstar/blob/master/src/mol-plugin-state/helpers/structure-overpaint.ts#L21) (or just get the required code from it's body)
  - The overpaint will add new nodes to the state tree
  - To get get structure `StructureComponentRef`, you just remove the `?.cell` in the `data` variable

**Reply by Manubi** [reply]
Thanks for the fast and great support!

**Reply by Manubi** [reply]
Another question. What's the easiest way of changing the color of a selection? Thanks again
 ```ts
 async function selectResidue(selectedResidue: Set<number>) {
    const data =
      window.molstar?.managers.structure.hierarchy.current.structures[0]?.cell
        .obj?.data;
    if (!data) return;
    window.molstar?.managers.interactivity.lociSelects.deselectAll();

    for (const seq_id of selectedResidue) {
      const sel = Script.getStructureSelection(
        (Q) =>
          Q.struct.generator.atomGroups({
            "residue-test": Q.core.rel.eq([
              Q.struct.atomProperty.macromolecular.label_seq_id(),
              seq_id,
            ]),
            "group-by": Q.struct.atomProperty.macromolecular.residueKey(),
          }),
        data
      );
      const loci = StructureSelection.toLociWithSourceUnits(sel);
      window.molstar?.managers.interactivity.lociSelects.select({
        loci,
      });
    }
  }```

**Reply by dsehnal** [reply]
The overpaint is a general mechanism to achieve that. Otherwise you would have to write your own color theme and store the selections and colors in the params, which is somewhat more complicated.

Mol* doesn't have a concept for "selection color" at the basic data model level.
